add end points for carts and implement middleware for authentication

AntonioLok opened the pull request:
# Description

This PR implements the follow:
1 - Add routes for getting the associate cart for an authenticated user and a route for updated
2 - Add authentication middleware which checks for the token, and verify that the token is not malformed and if its a valid token, then the user will have permission to access the 2 routes that I added above.

# Test
1 - Install dependencies 
2 - Set the required environment variable `JWT_SECRET` by creating a `.env` file in the main directory and set your secret in that file ie. `JWT_SECRET=HELLO`.
3 - run the app using `npm run start`. If you want to skip step 2, you can set the env with `JWT_SECRET=HELLO npm run start`.

Now, using the command line, send the following HTTP requests:

-  Register for an account first:
    -  method: POST http://localhost:8000/users/register
    -  body parameters:
        - `email` type: string
        - `password` type: string

sample command:
`curl -d "email=antonio@gmail.coml&password=password" -X POST http://localhost:8000/users/register`.

The response should be:

```
 {"statusCode":201,"statusText":"CREATED","data":null}
```
- Login with that same credential:
    -  method: POST http://localhost:8000/users/login
    -  body parameters:
        - `email` type: string
        - `password` type: string

Sample command:
`curl -d "email=antonio@gmail.coml&password=password" -X POST http://localhost:8000/users/login`

The response should be :

```
{"statusCode":200,"statusText":"SUCCESS","data":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImFudG9uaW9AZ21haWwuY29tbCIsImlhdCI6MTU2NzE5MjA4MCwiZXhwIjoxNTY3Mjc4NDgwfQ.KKh-7r2hpNHEbFD8RTQgMsucjGwGuyKaY7GXAfL8sgI"}
```
Note that the token was sent in the data field.

- Get the user cart:
    -  method: GET http://localhost:8000/carts
    -  body parameters: none

Sample command:
`curl -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImFudG9uaW9AZ21haWwuY29tbCIsImlhdCI6MTU2NzE5MjA4MCwiZXhwIjoxNTY3Mjc4NDgwfQ.KKh-7r2hpNHEbFD8RTQgMsucjGwGuyKaY7GXAfL8sgI" -X GET http://localhost:8000/carts`

Note that this request requires a bearer token to be passed so please pass the token that you got from the login response.

The response should be: 

```
{"statusCode":200,"statusText":"SUCCESS","data":[]}
```

As we can see from the response, the data is just en empty array as the cart is initially empty when a user creates a new account.

- Update the user cart:
    -  method: POST http://localhost:8000/carts
    -  body parameters: 
       '{"updatedProducts":  [{"_id": string, "size": string, "quantity": number}] }'

Sample command:
```
curl -X POST \
  http://localhost:8000/carts \
  -H 'Accept: */*' \
  -H 'Accept-Encoding: gzip, deflate' \
  -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImFudG9uaW9AZ21haWwuY29tbCIsImlhdCI6MTU2NzE5MjA4MCwiZXhwIjoxNTY3Mjc4NDgwfQ.KKh-7r2hpNHEbFD8RTQgMsucjGwGuyKaY7GXAfL8sgI' \
  -H 'Cache-Control: no-cache' \
  -H 'Connection: keep-alive' \
  -H 'Content-Length: 76' \
  -H 'Content-Type: application/json' \
  -H 'Host: localhost:8000' \
  -H 'Postman-Token: f891a9ca-5a73-460f-a314-ef49384c5222,75ba8bee-1a16-4bfa-9473-18d749635b8f' \
  -H 'User-Agent: PostmanRuntime/7.16.3' \
  -H 'cache-control: no-cache' \
  -d '{"updatedProducts":  [{"_id": "productId1", "size": "M", "quantity": 6}] }

'
```

Note that this request requires a bearer token to be passed so please pass the token that you got from the login response.

The response should be:

```
{"statusCode":200,"statusText":"SUCCESS","data":[{"_id":"productId1","size":"M","quantity":6}]}
```


# Next to do
Front end implementation for this

howardhenry commented:
> My apologies. I forgot to mention that you need to set the env variable for the secret. I had that already set in my machine and so when I re-did all the steps I did not have to do it and so I forgot to mention it.
> 
> You will have to set the env variable for JWT_SECRET
> 
> So when running the app:
> `JWT_SECRET=secret npm run start`

Ok.. got it.

A couple of points here:
1. There needs to be some instructions in the documentation that indicates that the `JWT_SECRET` env var is required to start the application.
2. You should provide a convenient mechanism to start the application with environment variables that are not hard coded into the npm scripts or anywhere else. dotenv (https://www.npmjs.com/package/dotenv) is a library that helps with that so you can check that out as a starting point.
3. I was going through your instructions to make the API requests and they weren't that easy to follow. You should generally avoid instructions that require a reviewer to download external applications (like Postman). The best set of instructions are those that leverage the command line as much as possible. In this case, it would be good to have all the instructions for API requests written out using `curl` as this is the defacto CLI tool for devs when testing HTTP requests.
4. The instructions didn't actually work for me. I got stuck on step #5 as when I made the request (based on what I understood from your instructions), I got back:
```
{
    "statusCode": 200,
    "statusText": "SUCCESS",
    "data": null
}
```
both after the POST and the subsequent GET request to _http://localhost:8000/carts_. 

In your PR description you need to outline the request structure and the expected response. I understand that you're trying to get the feature in quickly and learn about backend. And, I don't enjoy holding up your PR. However, the style in which you document what you have done is not standard and I find it difficult to read. I strongly recommend getting familiar with some of the common ways that API usage is documented. Take a look at some popular public APIs like MailChimp (https://developer.mailchimp.com/documentation/mailchimp/reference/campaigns/#read-get_campaigns), Github (https://developer.github.com/v3/interactions/repos/) etc. and start by following the same structure that they use to document the API usage. It helps tremendously in the PR process when a reviewer is identifying if there are gaps, bugs or inconsistencies with your implementation.

AntonioLok commented:
> > My apologies. I forgot to mention that you need to set the env variable for the secret. I had that already set in my machine and so when I re-did all the steps I did not have to do it and so I forgot to mention it.
> > You will have to set the env variable for JWT_SECRET
> > So when running the app:
> > `JWT_SECRET=secret npm run start`
> 
> Ok.. got it.
> 
> A couple of points here:
> 
> 1. There needs to be some instructions in the documentation that indicates that the `JWT_SECRET` env var is required to start the application.
> 2. You should provide a convenient mechanism to start the application with environment variables that are not hard coded into the npm scripts or anywhere else. dotenv (https://www.npmjs.com/package/dotenv) is a library that helps with that so you can check that out as a starting point.
> 3. I was going through your instructions to make the API requests and they weren't that easy to follow. You should generally avoid instructions that require a reviewer to download external applications (like Postman). The best set of instructions are those that leverage the command line as much as possible. In this case, it would be good to have all the instructions for API requests written out using `curl` as this is the defacto CLI tool for devs when testing HTTP requests.
> 4. The instructions didn't actually work for me. I got stuck on step #5 as when I made the request (based on what I understood from your instructions), I got back:
> 
> ```
> {
>     "statusCode": 200,
>     "statusText": "SUCCESS",
>     "data": null
> }
> ```
> 
> both after the POST and the subsequent GET request to _http://localhost:8000/carts_.
> 
> In your PR description you need to outline the request structure and the expected response. I understand that you're trying to get the feature in quickly and learn about backend. And, I don't enjoy holding up your PR. However, the style in which you document what you have done is not standard and I find it difficult to read. I strongly recommend getting familiar with some of the common ways that API usage is documented. Take a look at some popular public APIs like MailChimp (https://developer.mailchimp.com/documentation/mailchimp/reference/campaigns/#read-get_campaigns), Github (https://developer.github.com/v3/interactions/repos/) etc. and start by following the same structure that they use to document the API usage. It helps tremendously in the PR process when a reviewer is identifying if there are gaps, bugs or inconsistencies with your implementation.

Hey, really sorry about all of this. I have (hopefully) addressed all your comments. Hopefully from now own my PR description is much better which can help both of us review accept the Prs. I apologize for the troubles, it was not my intention to not make it harder for you to review. Please let me know if its good now.

howardhenry commented:
Looks good! Instructions were much more clear and they worked this time 👍

AntonioLok commented on src/controllers/user/register.js:
```diff
@@ -12,8 +13,10 @@ const register = async (username, password) => {
+    const newCart = new Cart({ username: newUser.username });
 
     await newUser.save();
+    await newCart.save();
```
Yup, user can only have cart. I am not sure why would a user have multiple cart... I dont shop online that much but I usually see one big cart where users can dump bunch of products in there.

Sample cart object:

```
cart = {
  username: 'antonio@gmail.com`,
  products = {
    productId1: {
      'S': 3
      'M': 4
    },
    productId2: {
      'S': 3
      'M': 4
    },
    productId3: {
      'L': 1
      'M': 1
    },
  }
}
```

S, M, L are the sizes of the product. They are used as key and their values indicate how many quantities for that selected size. How is this clear!

howardhenry replied:
Ok cool. It makes sense.

I'm seeing that retrieving the cart contents could however be tedious. Let's look at the following case:
* I want to know how many items are in a user's cart. Here you have 6 items, but there's no easy way to simple do a count programmatically. Based on your current structure you would have to do a recursive `lodash` forEach. Instead, consider the following structure:
```
cart = {
  username: 'antonio@gmail.com',
  products: [
    { id: productId1, quantity: 3, size: 'small', dateAdded: '2019-09-01' },
    { id: productId1, quantity: 4, size: 'medium' },
    { id: productId2, quantity: 3, size: 'small' },
    { id: productId2, quantity: 4 size: 'medium' },
    { id: productId3, quantity: 1, size: 'large' },
    { id: productId3, quantity: 1, size: 'medium' }
  ]
}
```
This way, you can simply do a `count` of the product array to know the total in your cart. You can also do `lodash.groupBy` or `lodash.sortBy` to group or sort by id or date or whatever field you need to. Basically, the goal is to have a versatile data structure that allows your frontend to easily transform the data in the way it needs.

AntonioLok replied:
1. Yeah, I originally wanted to choose arrays but decided to just use objects. I am not a huge fan of arrays approach. Mainly because arrays have a linear time complexity for adding/removing vs objects/map having a constant constant time complexity for adding/removing https://medium.com/@sherryhsu/js-objects-and-arrays-which-one-is-faster-cfcdb1281704.

```
  products: [
    { id: productId1, quantity: 3, size: 'small', dateAdded: '2019-09-01' },
    { id: productId1, quantity: 4, size: 'medium' },
    { id: productId2, quantity: 3, size: 'small' },
    { id: productId2, quantity: 4 size: 'medium' },
    { id: productId3, quantity: 1, size: 'large' },
    { id: productId3, quantity: 1, size: 'medium' }
  ]
```

Suppose a user wants to update this item on their cart from
```    { id: productId3, quantity: 1, size: 'large' },```.

to

```    { id: productId3, quantity: 2, size: 'large' },```.
We would need to loop through the entire array which would take O(n) to find that item and then update the quantity field. But with objects, we don't need to loop and updating a field should take O(1).

_______
 
2. "Basically, the goal is to have a versatile data structure that allows your frontend to easily transform the data in the way it needs."

While this is true, we also need to consider the time complexity that some algorithms take in order to achieve what we want. If the cart arrays grow big, iterating through each items to remove/update an item will slow down our application.

AntonioLok replied:
Having said that.... calculating the total number of items in the cart with my approach would take a long time due to having recursive calls so thats a big down for my approach. I'm unsure now about which approach is better. I will keep thinking about it :/

howardhenry replied:
1. Accessing properties in objects is in fact faster than in an array. However, that doesn't fit any of your use cases. t's good to understand theoretically what can happen but it only applies if in practice there is a real difference. There is only a margin difference in speed with the two approaches, which you will not see even at a large scale. Separately, I can't think of a case where you would need to select a single item (by id) from the cart.

2. The benefit here that I'm suggesting isn't to optimize for speed, it's to optimize for readability. Developers reading a recursive loop will find that it is more complex than a flat loop. So, code maintainability improves with the array approach as it is easier to handle the data when the structure is simpler.

AntonioLok commented on src/models/cart.js:
```diff
@@ -0,0 +1,24 @@
+  // object of dynamic keys. As shown in the sample below, 1 is the productId.
+  // Cannot define dynamic keys in mongo schema
+  products: { type: Object, default: {} },
+}, { minimize: false }); // {1: {sizeSelected: 's', amount: 3 }, {sizeSelected: 'm', amount: 6 },..},
```
Oh no, these comments were made by me. I did not copy paste anything here. The comment: 
```// {1: {sizeSelected: 's', amount: 3 }, {sizeSelected: 'm', amount: 6 },..},``` was kinda misplaced. It was meant to describe how the products field within a cart will look like. 

I was just wanted to highlight how the products object will look like by giving an example
```
cart = {
  username: 'antonio@gmail.com`,
  products = {
    productId1: {
      'S': 3
      'M': 4
    },
    productId2: {
      'S': 3
      'M': 4
    },
    productId3: {
      'L': 1
      'M': 1
    },
  }
}
```

I will update the comments to make it look nicer.

howardhenry replied:
Ah ok. I understand why you made the comments now. 

However, where possible, you should try to have your code be self-documenting. And, an object model is the perfect case for that. In fact, the object model is documentation for what kind of object a particular model is expecting. My previous comment regarding using an array instead of object literals for the products makes it easier to define a self-documenting object model. So, it would look something like this...

```
const cartSchema = mongoose.Schema({
  username: { type: String, required: true },
  products: [{
    id: { type: String, required: true },
    size: { type: String, enum: ['small', 'medium', 'large'] },
    quantity: { type: Integer, required: true }
  }],
}, { minimize: false });
```

AntonioLok commented on src/models/cart.js:
```diff
@@ -0,0 +1,24 @@
+}, { minimize: false }); // {1: {sizeSelected: 's', amount: 3 }, {sizeSelected: 'm', amount: 6 },..},
+
+cartSchema.statics.getCart = async (username) => {
+  const cart = await cartModel.findOne({ username }).exec();
```
Now that I think about it, do we need the await in this case? I was updating this and got 
```Redundant use of `await` on a return value```

AntonioLok commented on src/models/cart.js:
```diff
@@ -0,0 +1,24 @@
+  return cart;
+};
+
+cartSchema.statics.updateCart = async (username, updatedProducts) => {
```
Await is really not needed in all these statements. 

The places where I use this function has the await. I believe its fine that its returning a promise... the await is handled in the actual controller. For example, in controllers/update-cart.js. We have the await for this function `    return await Cart.updateCart(userId, updatedProducts);` which will wait for this promise to resolve.  

Okay this is just a question... shouldn't the `try/catch` happen on the block of code that is using this function? I feel like it might be too repetitive adding try/catch everything. Like lets look at controllers/update-cart.js again. There is a `try/catch` there and adding another `try/catch` in this function `cartSchema.statics.updateCart = async (username, updatedProducts) => {` seem kinda redundant IMO. But I would like to hear your opinion

howardhenry replied:
You're absolutely correct. `async` functions are themselves promises so it is redundant to use `return await <some-method>` within an function. You can update the code to simple return the promise `return Cart.updateCart(userId, updatedProducts)`

Ideally, every method should be wrapped in a `try/catch` block _unless_ you want to defer the error handling up to a method that is earlier in the call chain. The benefit of wrapping all methods in a `try/catch` is that when a method throws an error you stack trace will indicate exactly which function and line threw the error. If it is deferred to the calling method, the stack trace will show that the calling method threw the error, which is not always the case.

All that being said, when an error is thrown, JavaScript stores all calling methods in the stack trace so you could technically read/parse that to find where your error is sourced. However, this is tedious and requires you printing and reading the entire stack trace instead of a one-line error message which includes the filepath, function and line number where the error originated.

AntonioLok commented on src/utils/middlewares/get-token-from-header.js:
```diff
@@ -0,0 +1,10 @@
+const getTokenFromheader = (req, res, next) => {
+  const header = req.headers.authorization;
+  const bearer = header.split(' ');
```
Thanks for the tip! 

Regarding... `Separately, wrap your code in a try catch. You don't want an exception crashing your app.` I will address this after we have come to an agreement based on a point I raised in some previous comment about adding a `try/catch` inside the actual function vs the block of code that uses that function

AntonioLok commented:
@howardhenry Thank you for the comments. I've updated my PR accordingly. Sorry for the delay... it took me quite a while to figure out how to send array data in the body field. It was being converted to a string.... Hopefully everything is good. Thanks for your time!

AntonioLok commented:
> Looks great! Nice job 👍

Thank you!
